### PR TITLE
Bugfix/download only response

### LINKED ISSUE #159: downloadonly deployment is never finished
When `downloadonly` assignment is pushed from Hawkbit, it's finished with the action status `none`. According to [DDI documentation](https://eclipse.dev/hawkbit/rest-api/rootcontroller-api-guide.html#_post_tenantcontrollerv1controlleriddeploymentbaseactionidfeedback) this would keep action in the state `running` and is never finished (which can be observed in the Hawkbit Management UI). But `downloadonly` should not expect any other activities after download, so we should finish the action after download is complete.

### PR DESCRIPTION
**Pull request by UVV-gh**
Modify unit test and response status for `downloadonly` deployments

Fixes #159

### CODE REVIEW

File: src/hawkbit-client.c [hunk 1]
```diff
@@ -914,6 +914,17 @@ static gpointer download_thread(gpointer data)
         }
 
         g_mutex_lock(&active_action->mutex);
+        // skip installation if hawkBit asked us to do so
+        if (!artifact->do_install &&
+            (!artifact->do_maintenance || g_strcmp0(artifact->do_maintenance, "available") == 0)) {
```
Comment (Bastian-Krause) [inline src/hawkbit-client.c:919]: How is this related to the maintenance window? Does [downloadonly](https://eclipse.dev/hawkbit/rest-api/targets-api-guide.html#_request_fields_5) mean `deployment.update = skip` and `deployment.maintenanceWindow = available` (or undefined)?

File: src/hawkbit-client.c [hunk 2]
```diff
@@ -914,6 +914,17 @@ static gpointer download_thread(gpointer data)
...
+                active_action->state = ACTION_STATE_SUCCESS;
```
Comment (Bastian-Krause) [inline src/hawkbit-client.c:920]: Does this change anything, really?

File: src/hawkbit-client.c [hunk 3]
```diff
@@ -914,6 +914,17 @@ static gpointer download_thread(gpointer data)
...
+                if (!feedback(artifact->feedback_url, active_action->id, "File checksum OK.", "success",
+                              "downloaded", &feedback_error))
+                        g_warning("%s", feedback_error->message);
```
Comment (Bastian-Krause) [inline src/hawkbit-client.c:923]: Wouldn't this also apply to the "normal code path" in this function? So rather replace the `if !feedback_progress(artifact->feedback_url, active_action->id, "File checksum OK.", &error)) {` below?

File: src/hawkbit-client.c [hunk 4]
```diff
@@ -934,7 +945,6 @@ static gpointer download_thread(gpointer data)
 
                 return GINT_TO_POINTER(TRUE);
         }
-
```
Comment (Bastian-Krause) [inline src/hawkbit-client.c:937]: Seems unrelated.

File: include/hawkbit-client.h [hunk 5]
```diff
@@ -90,6 +90,7 @@ typedef struct Artifact_ {
         gchar *download_url;          /**< download URL of software bundle file */
         gchar *feedback_url;          /**< URL status feedback should be sent to */
         gchar *sha1;                  /**< sha1 checksum of software bundle file */
+        gchar *do_maintenance;        /**< maintenance flag, possible values: available, unavailable, null */
```
Comment (Bastian-Krause) [inline include/hawkbit-client.h:93]: `do_maintenance` sounds like a boolean. Maybe `maintenance_window` is better and closer to the DDI API?

**Review by Bastian-Krause — COMMENTED** [review]
How do you use `downloadonly`? If you use the downloaded RAUC bundle outside of rauc-hawkbit-updater, how can you be sure that it won't be replaced by another download while you're using it?

File: src/hawkbit-client.c (hunk 1 above)
Comment (UVV-gh) [reply src/hawkbit-client.c:919]: Maintenance is communicated with the same DDI parameters (force + skip) . See the corresponding test case for [download only](https://github.com/eclipse/hawkbit/blob/593a0bb146ff49442c78e786ab969b28b599d81a/hawkbit-rest/hawkbit-ddi-resource/src/test/java/org/eclipse/hawkbit/ddi/rest/resource/DdiDeploymentBaseTest.java#L348) and [maintenance](https://github.com/eclipse/hawkbit/blob/593a0bb146ff49442c78e786ab969b28b599d81a/hawkbit-rest/hawkbit-ddi-resource/src/test/java/org/eclipse/hawkbit/ddi/rest/resource/DdiRootControllerTest.java#L654)
`downloadonly` would have maintenance window `undefined` in this case

File: include/hawkbit-client.h (hunk 5 above)
Comment (UVV-gh) [reply include/hawkbit-client.h:93]: It's actually 3 state like mentioned in the comments, `undefined` does not mean `false`. I'm OK to rename, but wouldn't it be confusing with Target API, which actually defines the window (not just availability).

File: src/hawkbit-client.c (hunk 2 above)
Comment (UVV-gh) [reply src/hawkbit-client.c:920]: Good question. You mean that feedback alone would suffice?

File: src/hawkbit-client.c (hunk 3 above)
Comment (UVV-gh) [reply src/hawkbit-client.c:923]: Normal code path reports `proceeding` status. Or I'm missing the point here?

**Comment by UVV-gh** [comment]
> How do you use `downloadonly`? If you use the downloaded RAUC bundle outside of rauc-hawkbit-updater, how can you be sure that it the download is complete and that it won't be replaced by another download while you're using it?

The backend implementation decides (with the user interaction) when to do the actual update after the bundle is already on the device. The next step is the implementation of `confirmationBase` end-point so a user could confirm the download+install step beforehand.

File: src/hawkbit-client.c (hunk 1 above)
Comment (Bastian-Krause) [reply src/hawkbit-client.c:919]: Ah I see.

File: src/hawkbit-client.c (hunk 3 above)
Comment (Bastian-Krause) [reply src/hawkbit-client.c:923]: Ah, I guess you're right. `success`/`downladed` only applies to the `downloadonly` case.

File: include/hawkbit-client.h (hunk 5 above)
Comment (Bastian-Krause) [reply include/hawkbit-client.h:93]: Hm, the Target API is part of the Management API which we don't implement anyway (apart from `script/hawkbit_mgmt.py`). So I'd opt for renaming it. The DDI API attribute calls it `deployment.maintenanceWindow`, so `maintenance_window` should be okay.

**Comment by UVV-gh** [comment]
> That's not clear to me, how is `downloadonly` involved here?

How would I put the bundle on the device without it? :)
This patch makes sure that download is complete.

But I think I understand your concern. What if another download is started while the bundle is installed outside of rauc-hawkbit-updater? I guess this falls a bit out of scope of this particular fix, but I do have a few thoughts on it:
- Make sure the rauc-hawkbit-updater is stopped when the bundle is installed.
- Enhance rauc-hawkbit-updater to use file name from the URL and not from config file (this might not resolve your original concern though)

**Comment by Bastian-Krause** [comment]
> > That's not clear to me, how is `downloadonly` involved here?
> 
> How would I put the bundle on the device without it? :) This patch makes sure that download is complete.
> 
> But I think I understand your concern. What if another download is started while the bundle is installed outside of rauc-hawkbit-updater?

Exactly.

> I guess this falls a bit out of scope of this particular fix, but I do have a few thoughts on it:
> 
> * Make sure the rauc-hawkbit-updater is stopped when the bundle is installed.

Instead of letting rauc-hawkbit-updater run as a daemon, you could probably call `rauc-hawkbit-updater --run-once` every once in a while and check if a bundle appeared at the download location after it terminated.

> * Enhance rauc-hawkbit-updater to use file name from the URL and not from config file (this might not resolve your original concern though)

File: src/hawkbit-client.c (hunk 4 above)
Comment (UVV-gh) [reply src/hawkbit-client.c]: Fixed

File: include/hawkbit-client.h (hunk 5 above)
Comment (UVV-gh) [reply include/hawkbit-client.h]: Done

File: include/hawkbit-client.h [hunk 6]
```diff
@@ -90,6 +90,7 @@ typedef struct Artifact_ {
         gchar *download_url;          /**< download URL of software bundle file */
         gchar *feedback_url;          /**< URL status feedback should be sent to */
         gchar *sha1;                  /**< sha1 checksum of software bundle file */
+        gchar *maintenance_window;        /**< maintenance flag, possible values: available, unavailable, null */
```
Comment (Bastian-Krause) [inline include/hawkbit-client.h:93]:
```suggestion
        gchar *maintenance_window;    /**< maintenance flag, possible values: available, unavailable, null */
```
Nit pick..

**Review by Bastian-Krause — COMMENTED** [review]
Everything else looks good.

File: include/hawkbit-client.h (hunk 6 above)
Comment (UVV-gh) [reply include/hawkbit-client.h]: Overlooked after auto-renaming :) Fixed

**Review by Bastian-Krause — APPROVED** [review]
Thanks!
